src/accountsdb/snapshots.zig

```diff
@@ -1052,22 +1372,129 @@ pub const AccountsDbFields = struct {
     slot: Slot,
     bank_hash_info: BankHashInfo,
 
-    // default on EOF
     /// NOTE: these are currently always empty?
-    /// https://github.com/anza-xyz/agave/blob/b9eb4e2aa328abb9d3ee1d857d82ccd7a86f8c4d/runtime/src/serde_snapshot.rs#L769-L782
-    rooted_slots: std.ArrayListUnmanaged(Slot),
-    rooted_slot_hashes: std.ArrayListUnmanaged(SlotAndHash),
-
-    pub const @"!bincode-config:file_map" = bincode.hashmap.hashMapFieldConfig(FileMap, .{
-        .value = bincode.list.valueEncodedAsSlice(AccountFileInfo, .{}),
-    });
-    pub const @"!bincode-config:rooted_slots" = bincode.arraylist.defaultOnEofConfig(std.ArrayListUnmanaged(Slot));
-    pub const @"!bincode-config:rooted_slot_hashes" = bincode.arraylist.defaultOnEofConfig(std.ArrayListUnmanaged(SlotAndHash));
+    /// https://github.com/anza-xyz/agave/blob/9c899a72414993dc005f11afb5df10752b10810b/runtime/src/serde_snapshot.rs#L815-L825
+    rooted_slots: []const Slot,
+    rooted_slot_hashes: []const SlotAndHash,
 
-    pub const FileMap = std.AutoArrayHashMap(Slot, AccountFileInfo);
+    pub const FileMap = std.AutoArrayHashMapUnmanaged(Slot, AccountFileInfo);
 
     pub fn deinit(fields: AccountsDbFields, allocator: std.mem.Allocator) void {
-        bincode.free(allocator, fields);
+        var file_map = fields.file_map;
+        file_map.deinit(allocator);
+
+        allocator.free(fields.rooted_slots);
+        allocator.free(fields.rooted_slot_hashes);
+    }
+
+    pub const @"!bincode-config": bincode.FieldConfig(AccountsDbFields) = .{
+        .deserializer = bincodeRead,
+        .serializer = bincodeWrite,
+        .free = bincodeFree,
+    };
+
+    fn bincodeRead(
+        allocator: std.mem.Allocator,
+        reader: anytype,
+        params: bincode.Params,
+    ) !AccountsDbFields {
+        const assert_allocator = sig.utils.allocators.failing.allocator(.{
+            .alloc = .assert,
+            .resize = .assert,
+            .free = .assert,
+        });
+
+        var file_map = blk: {
+            const file_map_len: usize = try bincode.readIntAsLength(usize, reader, params) orelse
+                return error.FileMapTooBig;
+            var file_map: FileMap = .{};
+            errdefer file_map.deinit(allocator);
+
+            try file_map.ensureTotalCapacity(allocator, file_map_len);
+
+            for (0..file_map_len) |_| {
+                const key = try bincode.readInt(Slot, reader, params);
+
+                if (try bincode.readIntAsLength(usize, reader, params) != 1) {
+                    return error.TooManyAccountFileInfos;
+                }
+                const value = try bincode.read(assert_allocator, AccountFileInfo, reader, params);
+
+                const gop = file_map.getOrPutAssumeCapacity(key);
+                if (gop.found_existing) return error.DuplicateFileMapEntry;
+                gop.value_ptr.* = value;
+            }
+
+            break :blk file_map;
+        };
```

Comment:
why couldnt we just call bincode.read(Hashmap(Slot, []AccountFileInfo), ...) ? - feels like it would be better to have this as a per-field config as it was before?

Reply:
Well, reading it as a `HashMap(Slot, []AccountFileInfo)` directly would require an extra allocating conversion, so that would be less than ideal, this mimics the behavior of the original field config.

The per-field config approach was what originally hid the incorrect deserialization via the default-on-eof config, which I'm trying to avoid.

If the concern is the amount of logic here, I would suggest instead adding a simple function that does the equivalent, to be called here.
For example:
```zig
var file_map = try bincode.hashmap.readCtx(allocator, FileMap, params, struct {
    pub const readKey = {};
    pub fn readValue(_allocator: std.mem.Allocator, _reader: anytype, _params: bincode.Params) !AccountFileInfo {
        if (try bincode.readIntAsLength(usize, reader, params) != 1) {
            return error.TooManyAccountFileInfos;
        }
        return try bincode.read(_allocator, AccountFileInfo, _reader, _params);
    }
});
```
What do you think of that?

Reply:
nice - yeah im good with that - we would also want to remove `hashMapFieldConfig` too right? since it would be deadcode

Reply:
For the time being it isn't feasible to delete that just yet, it's still the basis of the [de]serialization of a lot of code we'll have to chip away at.

Reply:
ca28a947b0176442e8c44995f87a9f0156e239b2